### PR TITLE
add donwload command line and local_dir parameter

### PR DESCRIPTION
**Pull request by liuyhwangyh**
1. Download api support local_dir.
2. add download command
3. add login command

### CODE REVIEW

File: modelscope/hub/snapshot_download.py [hunk 1]
```diff
@@ -50,6 +53,9 @@ def snapshot_download(
         cookies (CookieJar, optional): The cookie of the request, default None.
         ignore_file_pattern (`str` or `List`, *optional*, default to `None`):
             Any file pattern to be ignored in downloading, like exact file names or file extensions.
+        allow_file_pattern (`str` or `List`, *optional*, default to `None`):
+            Any file pattern to be downloading, like exact file names or file extensions.
+        local_dir (str, optional): Specific download file location.
     Raises:
```
Comment (yingdachen) [inline modelscope/hub/snapshot_download.py:59]: 同上

File: modelscope/hub/snapshot_download.py [hunk 2]
```diff
@@ -65,17 +71,21 @@ def snapshot_download(
...
-
-    if cache_dir is None:
-        cache_dir = get_model_cache_root()
-    if isinstance(cache_dir, Path):
-        cache_dir = str(cache_dir)
     group_or_owner, name = model_id_to_group_owner_name(model_id)
-    temporary_cache_dir = os.path.join(cache_dir, 'temp', group_or_owner, name)
-    os.makedirs(temporary_cache_dir, exist_ok=True)
-    name = name.replace('.', '___')
+    if local_dir is not None:
```
Comment (yingdachen) [inline modelscope/hub/snapshot_download.py:75]: snapshot download和file download到这两段逻辑有没有可能refactor到一起？

File: modelscope/hub/file_download.py [hunk 3]
```diff
@@ -74,14 +76,21 @@ def model_file_download(
...
     """
-    if cache_dir is None:
-        cache_dir = get_model_cache_root()
-    if isinstance(cache_dir, Path):
-        cache_dir = str(cache_dir)
     group_or_owner, name = model_id_to_group_owner_name(model_id)
-    temporary_cache_dir = os.path.join(cache_dir, 'temp', group_or_owner, name)
+    if local_dir is not None:
+        temporary_cache_dir = os.path.join(local_dir, TEMPORARY_FOLDER_NAME)
+        cache = ModelFileSystemCache(local_dir)
```
[the remark this answers is not in the document]
Comment (liuyhwangyh) [reply modelscope/hub/file_download.py:82]: 是的

File: modelscope/cli/download.py [hunk 4]
```diff
@@ -35,10 +39,56 @@ def define_args(parsers: ArgumentParser):
             type=str,
             default=None,
             help='Cache directory to save model.')
+        parser.add_argument(
+            '--local_dir',
+            type=str,
+            default=None,
+            help='The files in the library will be downloaded'
```
[the remark this answers is not in the document]
Comment (liuyhwangyh) [reply modelscope/cli/download.py]: updated

File: modelscope/hub/snapshot_download.py (hunk 1 above)
Comment (liuyhwangyh) [reply modelscope/hub/snapshot_download.py:56]: 更新了

File: modelscope/hub/snapshot_download.py (hunk 2 above)
Comment (liuyhwangyh) [reply modelscope/hub/snapshot_download.py]: 合并到一个函数了

File: modelscope/cli/download.py [hunk 5]
```diff
@@ -35,10 +39,56 @@ def define_args(parsers: ArgumentParser):
...
+            'directly in this directory without model id.'
+            'If this parameter provided cache_dir will be ignored.')
+        parser.add_argument(
+            'files',
+            type=str,
+            default=None,
+            nargs='*',
+            help='Specify file paths in the repository to download'
```
[the remark this answers is not in the document]
Comment (liuyhwangyh) [reply modelscope/cli/download.py]: 更新了

File: modelscope/cli/login.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+class LoginCMD(CLICommand):
+    name = 'login'
+
+    def __init__(self, args):
+        self.args = args
+
+    @staticmethod
+    def define_args(parsers: ArgumentParser):
+        """ define args for download command.
```
[the remark this answers is not in the document]
Comment (liuyhwangyh) [reply modelscope/cli/login.py]: 更新了

File: modelscope/hub/file_download.py [hunk 7]
```diff
@@ -55,6 +56,7 @@ def model_file_download(
         local_files_only (bool, optional):  If `True`, avoid downloading the file and return the path to the
             local cached file if it exists. if `False`, download the file anyway even it exists.
         cookies (CookieJar, optional): The cookie of download request.
+        local_dir (str, optional): Specific download file location.
```
[the remark this answers is not in the document]
Comment (liuyhwangyh) [reply modelscope/hub/file_download.py]: 更新了

**Review by yingdachen — APPROVED** [review]

File: modelscope/cli/download.py [hunk 8]
```diff
@@ -35,10 +39,55 @@ def define_args(parsers: ArgumentParser):
...
             default=None,
             help='Cache directory to save model.')
+        parser.add_argument(
+            '--local_dir',
+            type=str,
+            default=None,
+            help='File will be downloaded to local location specified by'
+            'local_dir, in this case, cache_dir parameter will be ignored.')
+        parser.add_argument(
+            'files',
```
Comment (wenmengzhou) [inline modelscope/cli/download.py:49]: --files?  does it support wildcard str like  a/* ?

Comment (liuyhwangyh) [reply modelscope/cli/download.py:49]: 支持

Comment (liuyhwangyh) [reply modelscope/cli/download.py:49]: 和include效果一样

File: modelscope/cli/download.py [hunk 9]
```diff
@@ -35,10 +39,55 @@ def define_args(parsers: ArgumentParser):
...
     def execute(self):
-        snapshot_download(
-            self.args.model,
-            cache_dir=self.args.cache_dir,
-            revision=self.args.revision)
+        if len(self.args.files) == 1:  # download single file
+            model_file_download(
+                self.args.model,
+                self.args.files[0],
+                revision=self.args.revision)
```
[the remark this answers is not in the document]
Comment (liuyhwangyh) [reply modelscope/cli/download.py:76]: fixed.

File: tests/cli/test_download_cmd.py [hunk 10]
```diff
@@ -53,12 +53,12 @@ def tearDown(self):
         super().tearDown()
 
     def test_download(self):
-        cmd = f'python -m modelscope.cli.cli download {self.model_id}'
```
Comment (wenmengzhou) [inline tests/cli/test_download_cmd.py:56]: add test case for downloaing files to local_dir

File: modelscope/cli/download.py [hunk 11]
```diff
@@ -35,10 +39,57 @@ def define_args(parsers: ArgumentParser):
...
+            '--exclude',
+            nargs='*',
+            type=str,
+            default=None,
+            help='Glob patterns to exclude from files to download.'
+            'Ignored if file is specified')
         parser.set_defaults(func=subparser_func)
 
     def execute(self):
-        snapshot_download(
```
Comment (wenmengzhou) [inline modelscope/cli/download.py:41]: check only one of the two parameters is set:  cache_dir and local_dir

**Pull request merged**